sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -137,8 +138,16 @@ case class EnsureRequirements(
         Some(finalCandidateSpecs.values.maxBy(_.numPartitions))
       }
 
+      // Check if 1) all children are of `KeyGroupedPartitioning` and 2) they are all compatible
+      // with each other. If both are true, skip shuffle.
+      val allCompatible = childrenIndexes.sliding(2).map {
+            case Seq(a, b) =>
+              checkKeyGroupedSpec(specs(a)) && checkKeyGroupedSpec(specs(b)) &&
```

Comment:
Do we need `checkKeyGroupedSpec`? It looks like we can always eliminate shuffles if all the children are compatible. We can update `KeyGroupedShuffleSpec.isCompatible` to always return false if the feature flag is off.

Reply:
makes sense